src/ore/src/tracing/optimizer.rs

```diff
@@ -0,0 +1,313 @@
+    pub duration: std::time::Duration,
+    /// Ancestor chain of span names (root is first, parent is last).
+    pub path: String,
+    /// The plan produced this step.
+    pub plan: T,
+}
+
+/// Trace a fragment of type `T` to be emitted as part of an `EXPLAIN OPTIMIZER
+/// TRACE` output.
+///
+/// For best compatibility with the existing UI (which at the moment is the only
+/// sane way to look at such `EXPLAIN` traces), code instrumentation should
+/// adhere to the following constraints:
+///
+/// 1.  The plan type should be listed in the layers created in the
+///     `OptimizerTrace` constructor.
+/// 2.  Each `trace_plan` should be unique within it's enclosing span and should
+///     represent the result of the stage idenified by that span. In particular,
+///     this means that functions that call `trace_plan` more than once need to
+///     construct ad-hoc spans (see the iteration spans in the `Fixpoint`
+///     transform for example).
+///
+/// As a consequence of the second constraint, a sequence of paths such as
+/// ```text
+/// optimizer.foo.bar
+/// optimizer.foo.baz
+/// ```
+/// is not well-formed as it is missing the results of the prefix paths at the
+/// end:
+/// ```text
+/// optimizer.foo.bar
+/// optimizer.foo.baz
+/// optimizer.foo
+/// optimizer
+/// ```
+///
+/// Also, note that full paths can be repeated within a pipeline, but adjacent
+/// duplicates are interpreted as separete invocations. For example, the
+/// sub-sequence
+/// ```text
+/// ... // preceding stages
+/// optimizer.foo.bar // 1st call
+/// optimizer.foo.bar // 2nd call
+/// ... // following stages
+/// ```
+/// will be rendered by the UI as the following tree structure.
+/// ```text
+/// optimizer
+///   ... // following stages
+///   foo
+///     bar // 2nd call
+///     bar // 1st call
+///   ... // preceding stages
+/// ```
+pub fn trace_plan<T: Clone + 'static>(plan: &T) {
+    tracing::Span::current().with_subscriber(|(_id, subscriber)| {
+        if let Some(trace) = subscriber.downcast_ref::<PlanTrace<T>>() {
+            // Take the current time before cloning the plan to exclude the clone()
+            // time from the duration computed by `push` when creating a new TraceEntry.
+            let current_time = std::time::Instant::now();
+            let plan = plan.clone();
+            trace.push(plan, current_time)
+        }
+    });
+}
+
+/// A [`layer::Layer`] implementation for [`PlanTrace`].
+///
+/// Populates the `data` wrapped by the [`PlanTrace`] instance with
+/// [`TraceEntry`] values, one for each span with attached plan in its
+/// extensions map.
+impl<S, T> layer::Layer<S> for PlanTrace<T>
+where
+    S: subscriber::Subscriber + for<'span> tracing_subscriber::registry::LookupSpan<'span>,
+    T: 'static,
+{
+    fn on_new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, ctx: layer::Context<'_, S>) {
+        // push to path stack
+        let mut path = self.path.lock().expect("path shouldn't be poisoned");
```

Comment:
how much slower is it when you have tracing on? One big change we could make here is using span `extensions_mut` to store the data instead: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/registry/trait.SpanData.html#tymethod.extensions_mut

I am not going to block this pr for this change, but i think it'd be worth it in the future. However, you might find that using a `RwLock` instead of a `Mutex` might be a perf improvement, depending on how often you are reading. (underneath, `ExtensionsMut` is using a `RwLock`)